### PR TITLE
Refactor distance calculations using cKDTree

### PR DESCRIPTION
**Pull request by kavanase**
This is a drop-in change to dramatically accelerate `pbc_coord_intersection` in `Structure.get_primitive_structure()`.
It accelerates `Structure.get_primitive_structure()` by about 10x (depending on the cell), and thus any usage of `StructureMatcher(...primitive_cell=True)` (default setting).

Timings test:
image

MWE:
```python
from pymatgen.core.structure import Structure

lmno = Structure.from_file("../tests/data/Li2Mn3NiO8_POSCAR")
lmno_supercell = lmno*3
for i in range(10):
    lmno_supercell.get_primitive_structure()  # 38.2 seconds with original code, 3.9s with update
```
Structure used:
[Li2Mn3NiO8_POSCAR.zip](https://github.com/user-attachments/files/27181634/Li2Mn3NiO8_POSCAR.zip)

### CODE REVIEW

File: src/pymatgen/core/structure.py [hunk 1]
```diff
@@ -2614,12 +2614,24 @@ def site_label(site):
...
+            scale = 1.0 / tol
+            boxsize = scale
+        
+            fc1_scaled = (fc1 % 1.0) * scale
+            fc2_scaled = (fc2 % 1.0) * scale
+        
+            # cKDTree requires all coords strictly < boxsize
+            np.clip(fc1_scaled, 0, boxsize * (1 - 1e-15), out=fc1_scaled)
+            np.clip(fc2_scaled, 0, boxsize * (1 - 1e-15), out=fc2_scaled)
+
```
[the remark this answers is not in the document]
Comment (kavanase) [reply src/pymatgen/core/structure.py]: Tol should always be non zero

**Comment by shyuep** [comment]
Nice! Can you add the additional tests for the new edge behavior as suggested by Co-Pilot?

**Comment by kavanase** [comment]
I can add the check for the edge case suggested by copilot, but it's referring to a case where `tolerance` has manually been set to 0 (or negative), which should never be the case  – if a user does this, an error _should_ be thrown?
If you'd still prefer adding the handling suggested by Copilot (to allow `tolerance=0`, where it should just (silently) fail to find any primitive structure), I'm happy to add, but just to check before adding any potentially unnecessary code.

**Comment by shyuep** [comment]
I understand. I think we can throw an error and check that

File: src/pymatgen/core/structure.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kavanase) [reply src/pymatgen/core/structure.py]: @Copilot `tolerance` (and thus `tol`) should never be <=0. Add a `ValueError` catch for this case, saying that `Structure.get_primitive_structure()` does not support `tolerance<=0`.

**Comment by kavanase** [comment]
Done!
